### PR TITLE
BigInteger based random testing of System.Decimal

### PR DESCRIPTION
**Pull request by pentp**
New comprehensive tests are needed for a significantly improved managed decimal implementation: https://github.com/dotnet/corert/pull/4452#issuecomment-328376473, at first in CoreRT, but hopefully later shared with CoreCLR.

These tests compare `System.Decimal` cmp/add/mul/div results against a `BigInteger` based reference implementation using ~1000 different random decimals covering every scale and sign with ~20 different bit-patterns each (about 950k combinations). It adds about 1-2 seconds to `System.Runtime` test execution time (from 10s to 12s) - is this acceptable?

### CODE REVIEW

**Comment by danmosemsft** [comment]
https://github.com/dotnet/corefx/issues/23960

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 1]
```diff
@@ -1293,5 +1295,375 @@ public static void DecrementOperator(decimal d, decimal expected)
...
+                if (high <= 32)
+                    res = 1u << (high - 1);
+                res |= (uint)Math.Ceiling((uint.MaxValue >> Math.Max(0, 32 - high)) * rnd.NextDouble());
+                if (low > 0)
+                    res = (res >> low) << low;
+                return (int)res;
+            }
+        }
+
+        struct BigDecimal
```
Comment (mellinoe) [inline src/System.Runtime/tests/System/DecimalTests.cs]: There's a very large amount of non-trivial code in this structure. Could you explain a bit more about where it came from / why it is correct? Is this logic copy-pasted from `BigInteger`? Without understanding how this stuff is valid / trustworthy, it's hard to put a lot of value in the tests that rely on it.

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 2]
```diff
@@ -1293,5 +1295,375 @@ public static void DecrementOperator(decimal d, decimal expected)
         {
             Assert.Equal(expected, --d);
         }
+
+        [Fact]
+        public static void RandomBigInteger_Compare()
+        {
+            var values = GetRandomData(out var bigValues);
+            for (int i = 0, b = 0; i < values.Length; i++)
```
Comment (mellinoe) [inline src/System.Runtime/tests/System/DecimalTests.cs]: I don't think this `b` variable is actually used (it's only incremented).

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 3]
```diff
@@ -1293,5 +1295,375 @@ public static void DecrementOperator(decimal d, decimal expected)
...
+            {
+                var d1 = values[i];
+                var b1 = bigValues[i];
+                for (int j = 0; j < values.Length; j++, b++)
+                {
+                    var d2 = values[j];
+                    var expected = b1.Add(bigValues[j], out var expectedOF);
+                    if (expectedOF)
+                    {
+                        if (--overflowBudget < 0)
```
Comment (mellinoe) [inline src/System.Runtime/tests/System/DecimalTests.cs]: What's this for?

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 4]
```diff
@@ -1293,5 +1295,375 @@ public static void DecrementOperator(decimal d, decimal expected)
...
+                }
+            }
+        }
+
+        [Fact]
+        public static void RandomBigInteger_Add()
+        {
+            var overflowBudget = 1000;
+            var values = GetRandomData(out var bigValues);
+            for (int i = 0, b = 0; i < values.Length; i++)
```
Comment (mellinoe) [inline src/System.Runtime/tests/System/DecimalTests.cs]: `b` is unused

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 5]
```diff
@@ -1293,5 +1295,375 @@ public static void DecrementOperator(decimal d, decimal expected)
...
+                }
+            }
+        }
+
+        [Fact]
+        public static void RandomBigInteger_Mul()
+        {
+            var overflowBudget = 1000;
+            var values = GetRandomData(out var bigValues);
+            for (int i = 0, b = 0; i < values.Length; i++)
```
Comment (mellinoe) [inline src/System.Runtime/tests/System/DecimalTests.cs]: `b` is unused.

File: src/System.Runtime/tests/System/DecimalTests.cs (hunk 3 above)
Comment (pentp) [reply src/System.Runtime/tests/System/DecimalTests.cs]: This is for limiting the number of overflow cases verified - without the limit there would be about 100k additional exceptions thrown (4-5s additional test execution time).

**Comment by pentp** [comment]
The `BigDecimal` implementation is roughly based on the oleaut32 native decimal code currently used by CoreCLR/CoreRT (especially `ScaleResult`), but optimized for simplicity instead of speed. I could remove a few shortcuts from the code, but I think it can't be made much simpler.

I added more comments for the tests and `BigDecimal` code. The correctness of `BigDecimal` is ensured by the very same tests because they are symmetrical – the results should be always the same for `System.Decimal` and `BigDecimal`, except maybe if in the future we decide to remove the compatibility quirks of `VarDecMul`.

**Comment by mellinoe** [comment]
> The correctness of BigDecimal is ensured by the very same tests because they are symmetrical – the results should be always the same for System.Decimal and BigDecimal

Well, it ensures that they have the same behavior, but not that they are correct. They could both have the same bug, and the tests would pass. I'm not saying that's what is happening, but by using the `BigDecimal` structure like this, we are implicitly trusting that it's implementation is bug-free. Looking at it again, there is not a TON of complexity in `BigDecimal`, but it is still several hundred lines of arithmetic code with lots of branching.

Could I see the source from which it is based on?

**Comment by pentp** [comment]
I agree, the tests ensure that the results are the same as today in CoreRT/CoreCLR/Desktop, but not that they are correct. The source is in [CoreRT](https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Decimal.DecCalc.cs#L496) or [CoreCLR/oleaut32](https://github.com/dotnet/coreclr/blob/master/src/classlibnative/bcltype/decimal.cpp#L1848) (it's the same thing basically, in C# and C++), but I've tried to use as little as possible from these.

I think I was a bit too hasty to conclude that `BigDecimal` can't be made much simpler, have to investigate a couple of approaches...

**Comment by pentp** [comment]
In fact, I did find a [bug in `VarDecDiv`](https://github.com/dotnet/corert/blob/703c124257880772db09cf605fbc1da563ee7636/src/System.Private.CoreLib/src/System/Decimal.DecCalc.cs#L2224) when doing optimization work on CoreRT - it looks like shift with carry is done here using `>` not `>=` to detect overflow, but it's very unlikely to be hit (`rgulRem[0]` must be exactly `0x80000000`). [Same line](https://github.com/dotnet/coreclr/blob/eae763affd9624fbddf1c30ead52d23651de98e4/src/palrt/decarith.cpp#L997) in original C++ code also has this bug.

And another [probable bug in `Remainder`](https://github.com/dotnet/coreclr/issues/12605), but this one looks like it's impossible to hit (but I haven't gotten around to proving that). This whole function is a good candidate for some major refactoring after the core decimal code is shared between CoreRT and CoreCLR.

**Comment by danmosemsft** [comment]
If the tests were split across two classes, I believe xunit may parallelize, potentially cutting the time taken.

@mellinoe please see responses above.

**Comment by pentp** [comment]
I simplified `BigDecimal.Div` significantly so it should be relatively straightforward now with only 4 branches.
Splitting the tests into separate classes did reduce the time taken by half.

**Comment by mellinoe** [comment]
Looks okay to me now. @joperezr Did you want to take a look?

**Comment by joperezr** [comment]
Merging now since the failing test in opensuse is not related (System.Drawing test)

**Pull request merged**